### ISSUE TITLE
[ CCE ] Add support for Gemma3 for cut cross entropy

### ISSUE THREAD

**Issue by NanoCode012** [axolotl-ai-cloud/axolotl#2438]
### ⚠️ Please check that this feature request hasn't been suggested before.

- [x] I searched previous [Ideas in Discussions](https://github.com/axolotl-ai-cloud/axolotl/discussions/categories/ideas) didn't find any similar feature requests.
- [x] I searched previous [Issues](https://github.com/axolotl-ai-cloud/axolotl/labels/enhancement) didn't find any similar feature requests.

### 🔖 Feature description

There's an upstream PR which has what we want (similar implementation with Gemma2, but with one extra param) https://github.com/apple/ml-cross-entropy/pull/25

### ✔️ Solution

We can wait or patch CCE ourselves incase this is not merged. Waiting for multimodal PR to get in first #2435 

### Acknowledgements

- [x] My issue title is concise, descriptive, and in title casing.
- [x] I have searched the existing issues to make sure this feature has not been requested yet.
- [x] I have provided enough information for the maintainers to understand and evaluate this request.

**Comment by NanoCode012** [comment]
@strangedove , I noticed you closed your upstream PR. Is there some issue? Otherwise, would you be interested in making the monkey patch here in Axolotl or we could make the change as well.

**Comment by strangedove** [comment]
Sorry I just saw this. 

> [@strangedove](https://github.com/strangedove) , I noticed you closed your upstream PR. Is there some issue? Otherwise, would you be interested in making the monkey patch here in Axolotl or we could make the change as well.

Basically yeah, soon after I tried to implement CCE you guys pushed your revision for multimodal models; this wound up breaking the Gemma 3 implementation I had running (it only worked for Gemma3ForCausalLM at the time). 

I've confirmed since yesterday that the revisions I made for Gemma3ForConditionalGeneration work; I'll be making another PR to CCE for Gemma3 in the next day or two after final checks. 

The PR for Cohere/Cohere2 seems to work fine, though; that wasn't impacted by the multimodal update. 

Edit: Can also submit it here if you can show me where it would need to go. I

**Comment by NanoCode012** [comment]
@strangedove , I re-checked your second PR to the repo. The patch I made is here https://github.com/axolotl-ai-cloud/axolotl/blob/main/src/axolotl/integrations/cut_cross_entropy/monkeypatch/gemma3.py (other model patches are in this folder too)

Initially I tested with your setup (thanks for the starter code btw :) )before I updated it for transformers v4.50.0 (num_logits deprecation etc).

For multi-modal mode, I added a new config to defer logits calculation to the ConditionalGeneration model instead https://github.com/axolotl-ai-cloud/axolotl/blob/0bfa180f7d90b1483b676b4dd4f6bf2caf36891b/src/axolotl/integrations/cut_cross_entropy/monkeypatch/gemma3.py#L59

There's also some small changes like supporting soft cap https://github.com/axolotl-ai-cloud/axolotl/blob/0bfa180f7d90b1483b676b4dd4f6bf2caf36891b/src/axolotl/integrations/cut_cross_entropy/monkeypatch/gemma3.py#L349

For cohere, I scaled the lm head weight by logit_scale to replace the logits scaling in its source https://github.com/axolotl-ai-cloud/axolotl/blob/0bfa180f7d90b1483b676b4dd4f6bf2caf36891b/src/axolotl/integrations/cut_cross_entropy/monkeypatch/cohere.py#L134

If any part of the implementation sounds off, please do let me know!

**Comment by strangedove** [comment]
Have you tested the patch with deepspeed? 

For Gemma3, the other main issue I had in my testing was when trying to use CCE with any deepspeed zero3 configuration (zero2 and others worked fine). The tied weights between the `lm_head` and `embed_tokens` didn't correctly gather the tensor before trying the linear cross-entropy calculations when called as `self.lm_head.weight`. 

On my end, replacing `self.lm_head.weight` or `self.language_model.model.lm_head.weight` in the `apply_lce()` function in the patch with `self.embed_tokens.weight` or `self.language_model.model.embed_tokens.weight` fixed this right up. 

(No problems with Cohere that I saw, that one has a separate lm_head tensor I believe so no issues with tied weights.)

Everything else looks fine at a glance, glad to see this being patched in - I know a lot of people who have been waiting for proper support for these models :)

**Comment by NanoCode012** [comment]
@strangedove , I didn't test with z3 at the time.

I just ran it against gemma3_text, gemma3 (text only and multimodal mode) and it seems to be running.

If you can repro that issue with current implementation, do let us know
